src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs

```diff
@@ -1101,63 +1121,124 @@ private static Symbol FindMostSpecificImplementation(Symbol interfaceMember, Nam
+                        implementations[i] = (methodSet, bases);
+                    }
+
+                    if (bases.ContainsKey(interfaceType))
+                    {
+                        // Previous candidate is more specific
+                        candidate = default;
                         break;
                     }
                 }
+
+                if (candidate.Count == 0)
+                {
+                    continue;
+                }
+
+                if (implementations.Count != 0)
+                {
+                    MultiDictionary<NamedTypeSymbol, NamedTypeSymbol> bases = interfaceType.InterfacesAndTheirBaseInterfacesWithDefinitionUseSiteDiagnostics(ref useSiteDiagnostics);
+
+                    for (int i = implementations.Count - 1; i >= 0; i--)
+                    {
+                        if (bases.ContainsKey(implementations[i].MethodSet.First().ContainingType))
+                        {
+                            // new candidate is more specific
+                            implementations.RemoveAt(i);
+                        }
+                    }
+
+                    implementations.Add((candidate, bases));
+                }
+                else
+                {
+                    implementations.Add((candidate, null));
+                }
             }
 
-            if (shadowedInterfaces != null)
+            Symbol result;
+
+            switch (implementations.Count)
             {
-                shadowedInterfaces.Free();
+                case 0:
+                    result = null;
+                    conflictingImplementation1 = null;
+                    conflictingImplementation2 = null;
+                    break;
+                case 1:
+                    MultiDictionary<Symbol, Symbol>.ValueSet methodSet = implementations[0].MethodSet;
+                    switch (methodSet.Count)
+                    {
+                        case 1:
+                            result = methodSet.Single();
+                            break;
+                        default:
+                            result = null;
+                            break;
+                    }
+
+                    conflictingImplementation1 = null;
+                    conflictingImplementation2 = null;
+                    break;
+                default:
+                    result = null;
+                    conflictingImplementation1 = implementations[0].MethodSet.First();
+                    conflictingImplementation2 = implementations[1].MethodSet.First();
+                    break;
             }
 
-            return implementation;
+            implementations.Free();
+            return result;
         }
 
-        private static Symbol FindImplementationInInterface(Symbol interfaceMember, NamedTypeSymbol interfaceType)
+        private static MultiDictionary<Symbol, Symbol>.ValueSet FindImplementationInInterface(Symbol interfaceMember, NamedTypeSymbol interfaceType)
```

Comment:
May want to add a doc comment here. It's taken me a while to puzzle out what the intent is here. Particularly around nullability.

Reply:
> May want to add a doc comment here. It's taken me a while to puzzle out what the intent is here. Particularly around nullability.

This change isn't really driven by nullability mismatch. This is a follow up on a recent fix in master for [https://github.com/dotnet/roslyn/issues/31974](https://github.com/dotnet/roslyn/issues/31974). That fix changed return type of the ```GetExplicitImplementationForInterfaceMember``` API to return a set of symbols instead of a single symbol, which made behavior more reasonable in presence of error situations (when there is more than one explicit implementation for the same member within the same type). That return type and more precise result is simply propagated here so that that knowledge isn't lost.

---
In reply to: [261700128](https://github.com/dotnet/roslyn/pull/33778#discussion_r261700128) [](ancestors = 261700128)